Make errors nicer when TDNR fails

runarorama opened the pull request:
Fixes #4632

This makes the error message for TDNR failures simpler and clearer.

* Removes references to the namespace, since we're usually in the project root and UCM tells you which namespace you're in anyway.
* Infers the likely cause a little better and doesn't suggest e.g. a totally missing symbol when there are TDNR suggestions available.
* Added "TDNR masking the true error" to the list of common causes.
* Removed defunct instructions for adding a library.
* Added clear instructions for resolving the error
* Suffixified the TDNR suggestions.
* Removed bullets from TDNR suggestion list.
* Improved wrapping.
* Sorted suggestions by the length of the FQN, as a proxy for relevance.

Example of new behaviour:

<img width="914" alt="image" src="https://github.com/unisonweb/unison/assets/130916/2681ef35-5593-4757-a073-51644116181b">

runarorama committed:
Make errors nicer when TDNR fails

runarorama committed:
Merge branch 'trunk' of https://github.com/unisonweb/unison into runarorama/improve-tdnr-message

runarorama committed:
Sort by name length as a proxy for relevance

runarorama committed:
Transcript output

runarorama committed:
Rewrites output changed

ChrisPenner commented on parser-typechecker/src/Unison/PrintError.hs:
```diff
@@ -173,7 +173,7 @@ renderTypeError ::
   Path.Absolute ->
   Pretty ColorText
-renderTypeError e env src curPath = case e of
+renderTypeError e env src _curPath = case e of
```
Are you able to remove this arg entirely please?

ChrisPenner commented on parser-typechecker/src/Unison/PrintError.hs:
```diff
@@ -629,51 +629,68 @@ renderTypeError e env src curPath = case e of
     let (correct, wrongTypes, wrongNames) =
-          foldr sep id suggestions ([], [], [])
-        sep (C.Suggestion name typ _ match) r =
+          foldr sep id (sortOn (Text.length . C.suggestionName) suggestions) ([], [], [])
```
Is this meant to sort by the length of the FQN?
If you're aiming to select the most relevant names you might want to count the number of `.` separated segments instead.

ChrisPenner commented on parser-typechecker/src/Unison/PrintError.hs:
```diff
@@ -629,51 +629,68 @@ renderTypeError e env src curPath = case e of
+            C.Exact -> (_1 %~ (s :)) . r
+            C.WrongType -> (_2 %~ (s :)) . r
+            C.WrongName -> (_3 %~ (s :)) . r
+        -- libPath = Path.absoluteToPath' curPath Path.:> "lib"
```
```suggestion
```

ChrisPenner commented on unison-src/transcripts-manual/rewrites.output.md:
```diff
@@ -306,21 +306,17 @@ Instead, it should be an unbound free variable, which doesn't typecheck:
   
      19 |   bar21
   
+  Its type should conform to _ .
```
What's up with the type `_` here? Doesn't seem right 🤔

runarorama replied:
Hmm, used to be "there are no constraints on its type". I must have botched the pattern match on the type here.

ChrisPenner commented on unison-src/transcripts/destructuring-binds.output.md:
```diff
@@ -86,21 +86,17 @@ ex4 =
   
       2 |   (a,b) = (a Nat.+ b, 19)
   
+  Its type should conform to Nat .
```
Maybe either remove the `.` or drop the space :)

runarorama replied:
This is `wrap` adding spaces again

ChrisPenner reviewed (commented):
Looks good! I'm curious what's up with the transcripts outputting `_` as the expected type though 🤔

aryairani commented on parser-typechecker/src/Unison/PrintError.hs:
```diff
@@ -629,51 +629,68 @@ renderTypeError e env src curPath = case e of
+                          <> "\n\n"
+                          <> Pr.wrap "If that's not what you expected, you may have a type error somewhere else in your code."
+                    <> " "
+                    <> Pr.wrap (Pr.bold "Help me out by using a more specific name here or adding a type annotation.")
```
```suggestion
                    <> Pr.wrap ("Help me out by" <> Pr.bold "using a more specific name here" <> "or" <> Pr.bold "adding a type annotation.")
```
Maybe less bold, to draw attention more directly?

runarorama committed:
Update parser-typechecker/src/Unison/PrintError.hs

Co-authored-by: Chris Penner <christopher.penner@gmail.com>

runarorama committed:
Update parser-typechecker/src/Unison/PrintError.hs

Co-authored-by: Arya Irani <538571+aryairani@users.noreply.github.com>

runarorama committed:
Remove current path argument

runarorama committed:
Count segments as a proxy for relevance
instead of name length

runarorama committed:
Remove extraneous space

aryairani commented:
Currently a compile error in `parser-typechecker/tests/Unison/Test/UnisonSources.hs`

runarorama committed:
Fix up tests so they compile

runarorama committed:
Minor tweaks to output

runarorama committed:
Fix up tests

runarorama committed:
Merge branch 'trunk' of https://github.com/unisonweb/unison into runarorama/improve-tdnr-message

runarorama committed:
Incorporating latest formatter changes

runarorama committed:
I GUESS RUNNING IT AGAIN DID THE TRICK

runarorama committed:
Discard changes to unison-src/transcripts-manual/benchmarks/output/2024-01-25T15:48:31.296974000Z-bench.txt

aryairani commented:
> Looks good! I'm curious what's up with the transcripts outputting `_` as the expected type though 🤔

@runarorama I have this question too — It looks like it used to say 
> There are no constraints on its type.

but now it says 
> Its type should conform to _ 

(and missing the `.`)

Otherwise, I'm eager to get it merged before transcripts start failing again 😅

runarorama commented:
`Its type should conform to _` is a regression in trunk unrelated to this change.

I removed the `.` after the type because `Foo.Bar -> Baz.Qux` reads better than `Foo.Bar -> Baz.Qux.`

aryairani commented:
> `Its type should conform to _` is a regression in trunk unrelated to this change.

Hm, but I don't see that output in the transcripts on trunk?

https://github.com/unisonweb/unison/blob/trunk/unison-src/transcripts/resolution-failures.output.md?plain=1#L127
```
There are no constraints on its type.
```

vs

https://github.com/unisonweb/unison/blob/runarorama/improve-tdnr-message/unison-src/transcripts/resolution-failures.output.md?plain=1#L121
```
Its type should conform to:
  
      _
```

I think it might be due to losing [this logic](https://github.com/unisonweb/unison/pull/4641/files#diff-ea0df981901f1840e463f80adb91c04462f7f4e0b7978c159578e1467be88f60L658-L663), which is present [here](https://github.com/unisonweb/unison/pull/4641/files#diff-ea0df981901f1840e463f80adb91c04462f7f4e0b7978c159578e1467be88f60R679-R680), but missing [here](https://github.com/unisonweb/unison/pull/4641/files#diff-ea0df981901f1840e463f80adb91c04462f7f4e0b7978c159578e1467be88f60R651) and [here](https://github.com/unisonweb/unison/pull/4641/files#diff-ea0df981901f1840e463f80adb91c04462f7f4e0b7978c159578e1467be88f60R651).

> I removed the `.` after the type because `Foo.Bar -> Baz.Qux` reads better than `Foo.Bar -> Baz.Qux.`

Ok 👍

runarorama committed:
Fix type rendering for existentials, other tweaks

runarorama committed:
Merge branch 'runarorama/improve-tdnr-message' of https://github.com/unisonweb/unison into runarorama/improve-tdnr-message

runarorama commented:
OK, I believe I have addressed all of the issues and fixed the places where it was missing the existentials check. Output looks good in all transcripts. 🚢

aryairani reviewed (approved):
Great!
